// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/variants.proto

package com.google.genomics.v1;

/**
 * Protobuf type {@code google.genomics.v1.ImportVariantsRequest}
 *
 * <pre>
 * The variant data import request.
 * </pre>
 */
public  final class ImportVariantsRequest extends
    com.google.protobuf.GeneratedMessage implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.ImportVariantsRequest)
    ImportVariantsRequestOrBuilder {
  // Use ImportVariantsRequest.newBuilder() to construct.
  private ImportVariantsRequest(com.google.protobuf.GeneratedMessage.Builder builder) {
    super(builder);
  }
  private ImportVariantsRequest() {
    variantSetId_ = "";
    sourceUris_ = com.google.protobuf.LazyStringArrayList.EMPTY;
    format_ = 0;
    normalizeReferenceNames_ = false;
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
  }
  private ImportVariantsRequest(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
    this();
    int mutable_bitField0_ = 0;
    try {
      boolean done = false;
      while (!done) {
        int tag = input.readTag();
        switch (tag) {
          case 0:
            done = true;
            break;
          default: {
            if (!input.skipField(tag)) {
              done = true;
            }
            break;
          }
          case 10: {
            com.google.protobuf.ByteString bs = input.readBytes();

            variantSetId_ = bs;
            break;
          }
          case 18: {
            com.google.protobuf.ByteString bs = input.readBytes();
            if (!((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
              sourceUris_ = new com.google.protobuf.LazyStringArrayList();
              mutable_bitField0_ |= 0x00000002;
            }
            sourceUris_.add(bs);
            break;
          }
          case 24: {
            int rawValue = input.readEnum();

            format_ = rawValue;
            break;
          }
          case 40: {

            normalizeReferenceNames_ = input.readBool();
            break;
          }
        }
      }
    } catch (com.google.protobuf.InvalidProtocolBufferException e) {
      throw new RuntimeException(e.setUnfinishedMessage(this));
    } catch (java.io.IOException e) {
      throw new RuntimeException(
          new com.google.protobuf.InvalidProtocolBufferException(
              e.getMessage()).setUnfinishedMessage(this));
    } finally {
      if (((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
        sourceUris_ = sourceUris_.getUnmodifiableView();
      }
      makeExtensionsImmutable();
    }
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_ImportVariantsRequest_descriptor;
  }

  protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_ImportVariantsRequest_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.ImportVariantsRequest.class, com.google.genomics.v1.ImportVariantsRequest.Builder.class);
  }

  /**
   * Protobuf enum {@code google.genomics.v1.ImportVariantsRequest.Format}
   */
  public enum Format
      implements com.google.protobuf.ProtocolMessageEnum {
    /**
     * <code>FORMAT_UNSPECIFIED = 0;</code>
     */
    FORMAT_UNSPECIFIED(0, 0),
    /**
     * <code>FORMAT_VCF = 1;</code>
     *
     * <pre>
     * VCF (Variant Call Format). The VCF files should be uncompressed. gVCF is
     * also supported.
     * </pre>
     */
    FORMAT_VCF(1, 1),
    /**
     * <code>FORMAT_COMPLETE_GENOMICS = 2;</code>
     *
     * <pre>
     * Complete Genomics masterVarBeta format. The masterVarBeta files should
     * be bzip2 compressed.
     * </pre>
     */
    FORMAT_COMPLETE_GENOMICS(2, 2),
    UNRECOGNIZED(-1, -1),
    ;

    /**
     * <code>FORMAT_UNSPECIFIED = 0;</code>
     */
    public static final int FORMAT_UNSPECIFIED_VALUE = 0;
    /**
     * <code>FORMAT_VCF = 1;</code>
     *
     * <pre>
     * VCF (Variant Call Format). The VCF files should be uncompressed. gVCF is
     * also supported.
     * </pre>
     */
    public static final int FORMAT_VCF_VALUE = 1;
    /**
     * <code>FORMAT_COMPLETE_GENOMICS = 2;</code>
     *
     * <pre>
     * Complete Genomics masterVarBeta format. The masterVarBeta files should
     * be bzip2 compressed.
     * </pre>
     */
    public static final int FORMAT_COMPLETE_GENOMICS_VALUE = 2;


    public final int getNumber() {
      if (index == -1) {
        throw new java.lang.IllegalArgumentException(
            "Can't get the number of an unknown enum value.");
      }
      return value;
    }

    public static Format valueOf(int value) {
      switch (value) {
        case 0: return FORMAT_UNSPECIFIED;
        case 1: return FORMAT_VCF;
        case 2: return FORMAT_COMPLETE_GENOMICS;
        default: return null;
      }
    }

    public static com.google.protobuf.Internal.EnumLiteMap<Format>
        internalGetValueMap() {
      return internalValueMap;
    }
    private static com.google.protobuf.Internal.EnumLiteMap<Format>
        internalValueMap =
          new com.google.protobuf.Internal.EnumLiteMap<Format>() {
            public Format findValueByNumber(int number) {
              return Format.valueOf(number);
            }
          };

    public final com.google.protobuf.Descriptors.EnumValueDescriptor
        getValueDescriptor() {
      return getDescriptor().getValues().get(index);
    }
    public final com.google.protobuf.Descriptors.EnumDescriptor
        getDescriptorForType() {
      return getDescriptor();
    }
    public static final com.google.protobuf.Descriptors.EnumDescriptor
        getDescriptor() {
      return com.google.genomics.v1.ImportVariantsRequest.getDescriptor().getEnumTypes().get(0);
    }

    private static final Format[] VALUES = values();

    public static Format valueOf(
        com.google.protobuf.Descriptors.EnumValueDescriptor desc) {
      if (desc.getType() != getDescriptor()) {
        throw new java.lang.IllegalArgumentException(
          "EnumValueDescriptor is not for this type.");
      }
      if (desc.getIndex() == -1) {
        return UNRECOGNIZED;
      }
      return VALUES[desc.getIndex()];
    }

    private final int index;
    private final int value;

    private Format(int index, int value) {
      this.index = index;
      this.value = value;
    }

    // @@protoc_insertion_point(enum_scope:google.genomics.v1.ImportVariantsRequest.Format)
  }

  private int bitField0_;
  public static final int VARIANT_SET_ID_FIELD_NUMBER = 1;
  private volatile java.lang.Object variantSetId_;
  /**
   * <code>optional string variant_set_id = 1;</code>
   *
   * <pre>
   * Required. The variant set to which variant data should be imported.
   * </pre>
   */
  public java.lang.String getVariantSetId() {
    java.lang.Object ref = variantSetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      if (bs.isValidUtf8()) {
        variantSetId_ = s;
      }
      return s;
    }
  }
  /**
   * <code>optional string variant_set_id = 1;</code>
   *
   * <pre>
   * Required. The variant set to which variant data should be imported.
   * </pre>
   */
  public com.google.protobuf.ByteString
      getVariantSetIdBytes() {
    java.lang.Object ref = variantSetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      variantSetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int SOURCE_URIS_FIELD_NUMBER = 2;
  private com.google.protobuf.LazyStringList sourceUris_;
  /**
   * <code>repeated string source_uris = 2;</code>
   *
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   */
  public com.google.protobuf.ProtocolStringList
      getSourceUrisList() {
    return sourceUris_;
  }
  /**
   * <code>repeated string source_uris = 2;</code>
   *
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   */
  public int getSourceUrisCount() {
    return sourceUris_.size();
  }
  /**
   * <code>repeated string source_uris = 2;</code>
   *
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   */
  public java.lang.String getSourceUris(int index) {
    return sourceUris_.get(index);
  }
  /**
   * <code>repeated string source_uris = 2;</code>
   *
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   */
  public com.google.protobuf.ByteString
      getSourceUrisBytes(int index) {
    return sourceUris_.getByteString(index);
  }

  public static final int FORMAT_FIELD_NUMBER = 3;
  private int format_;
  /**
   * <code>optional .google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
   *
   * <pre>
   * The format of the variant data being imported. If unspecified, defaults to
   * to `VCF`.
   * </pre>
   */
  public int getFormatValue() {
    return format_;
  }
  /**
   * <code>optional .google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
   *
   * <pre>
   * The format of the variant data being imported. If unspecified, defaults to
   * to `VCF`.
   * </pre>
   */
  public com.google.genomics.v1.ImportVariantsRequest.Format getFormat() {
    com.google.genomics.v1.ImportVariantsRequest.Format result = com.google.genomics.v1.ImportVariantsRequest.Format.valueOf(format_);
    return result == null ? com.google.genomics.v1.ImportVariantsRequest.Format.UNRECOGNIZED : result;
  }

  public static final int NORMALIZE_REFERENCE_NAMES_FIELD_NUMBER = 5;
  private boolean normalizeReferenceNames_;
  /**
   * <code>optional bool normalize_reference_names = 5;</code>
   *
   * <pre>
   * Convert reference names to the canonical representation.
   * hg19 haploytypes (those reference names containing "_hap")
   * are not modified in any way.
   * All other reference names are modified according to the following rules:
   * The reference name is capitalized.
   * The "chr" prefix is dropped for all autosomes and sex chromsomes.
   * For example "chr17" becomes "17" and "chrX" becomes "X".
   * All mitochondrial chromosomes ("chrM", "chrMT", etc) become "MT".
   * </pre>
   */
  public boolean getNormalizeReferenceNames() {
    return normalizeReferenceNames_;
  }

  private byte memoizedIsInitialized = -1;
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (!getVariantSetIdBytes().isEmpty()) {
      output.writeBytes(1, getVariantSetIdBytes());
    }
    for (int i = 0; i < sourceUris_.size(); i++) {
      output.writeBytes(2, sourceUris_.getByteString(i));
    }
    if (format_ != com.google.genomics.v1.ImportVariantsRequest.Format.FORMAT_UNSPECIFIED.getNumber()) {
      output.writeEnum(3, format_);
    }
    if (normalizeReferenceNames_ != false) {
      output.writeBool(5, normalizeReferenceNames_);
    }
  }

  private int memoizedSerializedSize = -1;
  public int getSerializedSize() {
    int size = memoizedSerializedSize;
    if (size != -1) return size;

    size = 0;
    if (!getVariantSetIdBytes().isEmpty()) {
      size += com.google.protobuf.CodedOutputStream
        .computeBytesSize(1, getVariantSetIdBytes());
    }
    {
      int dataSize = 0;
      for (int i = 0; i < sourceUris_.size(); i++) {
        dataSize += com.google.protobuf.CodedOutputStream
          .computeBytesSizeNoTag(sourceUris_.getByteString(i));
      }
      size += dataSize;
      size += 1 * getSourceUrisList().size();
    }
    if (format_ != com.google.genomics.v1.ImportVariantsRequest.Format.FORMAT_UNSPECIFIED.getNumber()) {
      size += com.google.protobuf.CodedOutputStream
        .computeEnumSize(3, format_);
    }
    if (normalizeReferenceNames_ != false) {
      size += com.google.protobuf.CodedOutputStream
        .computeBoolSize(5, normalizeReferenceNames_);
    }
    memoizedSerializedSize = size;
    return size;
  }

  private static final long serialVersionUID = 0L;
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return PARSER.parseFrom(input);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return PARSER.parseFrom(input, extensionRegistry);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return PARSER.parseDelimitedFrom(input);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return PARSER.parseDelimitedFrom(input, extensionRegistry);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return PARSER.parseFrom(input);
  }
  public static com.google.genomics.v1.ImportVariantsRequest parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return PARSER.parseFrom(input, extensionRegistry);
  }

  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.ImportVariantsRequest prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessage.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * Protobuf type {@code google.genomics.v1.ImportVariantsRequest}
   *
   * <pre>
   * The variant data import request.
   * </pre>
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessage.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.ImportVariantsRequest)
      com.google.genomics.v1.ImportVariantsRequestOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_ImportVariantsRequest_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_ImportVariantsRequest_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.ImportVariantsRequest.class, com.google.genomics.v1.ImportVariantsRequest.Builder.class);
    }

    // Construct using com.google.genomics.v1.ImportVariantsRequest.newBuilder()
    private Builder() {
      maybeForceBuilderInitialization();
    }

    private Builder(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      super(parent);
      maybeForceBuilderInitialization();
    }
    private void maybeForceBuilderInitialization() {
      if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
      }
    }
    public Builder clear() {
      super.clear();
      variantSetId_ = "";

      sourceUris_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      bitField0_ = (bitField0_ & ~0x00000002);
      format_ = 0;

      normalizeReferenceNames_ = false;

      return this;
    }

    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_ImportVariantsRequest_descriptor;
    }

    public com.google.genomics.v1.ImportVariantsRequest getDefaultInstanceForType() {
      return com.google.genomics.v1.ImportVariantsRequest.getDefaultInstance();
    }

    public com.google.genomics.v1.ImportVariantsRequest build() {
      com.google.genomics.v1.ImportVariantsRequest result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    public com.google.genomics.v1.ImportVariantsRequest buildPartial() {
      com.google.genomics.v1.ImportVariantsRequest result = new com.google.genomics.v1.ImportVariantsRequest(this);
      int from_bitField0_ = bitField0_;
      int to_bitField0_ = 0;
      result.variantSetId_ = variantSetId_;
      if (((bitField0_ & 0x00000002) == 0x00000002)) {
        sourceUris_ = sourceUris_.getUnmodifiableView();
        bitField0_ = (bitField0_ & ~0x00000002);
      }
      result.sourceUris_ = sourceUris_;
      result.format_ = format_;
      result.normalizeReferenceNames_ = normalizeReferenceNames_;
      result.bitField0_ = to_bitField0_;
      onBuilt();
      return result;
    }

    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.ImportVariantsRequest) {
        return mergeFrom((com.google.genomics.v1.ImportVariantsRequest)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.ImportVariantsRequest other) {
      if (other == com.google.genomics.v1.ImportVariantsRequest.getDefaultInstance()) return this;
      if (!other.getVariantSetId().isEmpty()) {
        variantSetId_ = other.variantSetId_;
        onChanged();
      }
      if (!other.sourceUris_.isEmpty()) {
        if (sourceUris_.isEmpty()) {
          sourceUris_ = other.sourceUris_;
          bitField0_ = (bitField0_ & ~0x00000002);
        } else {
          ensureSourceUrisIsMutable();
          sourceUris_.addAll(other.sourceUris_);
        }
        onChanged();
      }
      if (other.format_ != 0) {
        setFormatValue(other.getFormatValue());
      }
      if (other.getNormalizeReferenceNames() != false) {
        setNormalizeReferenceNames(other.getNormalizeReferenceNames());
      }
      onChanged();
      return this;
    }

    public final boolean isInitialized() {
      return true;
    }

    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      com.google.genomics.v1.ImportVariantsRequest parsedMessage = null;
      try {
        parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        parsedMessage = (com.google.genomics.v1.ImportVariantsRequest) e.getUnfinishedMessage();
        throw e;
      } finally {
        if (parsedMessage != null) {
          mergeFrom(parsedMessage);
        }
      }
      return this;
    }
    private int bitField0_;

    private java.lang.Object variantSetId_ = "";
    /**
     * <code>optional string variant_set_id = 1;</code>
     *
     * <pre>
     * Required. The variant set to which variant data should be imported.
     * </pre>
     */
    public java.lang.String getVariantSetId() {
      java.lang.Object ref = variantSetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        if (bs.isValidUtf8()) {
          variantSetId_ = s;
        }
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <code>optional string variant_set_id = 1;</code>
     *
     * <pre>
     * Required. The variant set to which variant data should be imported.
     * </pre>
     */
    public com.google.protobuf.ByteString
        getVariantSetIdBytes() {
      java.lang.Object ref = variantSetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        variantSetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <code>optional string variant_set_id = 1;</code>
     *
     * <pre>
     * Required. The variant set to which variant data should be imported.
     * </pre>
     */
    public Builder setVariantSetId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      variantSetId_ = value;
      onChanged();
      return this;
    }
    /**
     * <code>optional string variant_set_id = 1;</code>
     *
     * <pre>
     * Required. The variant set to which variant data should be imported.
     * </pre>
     */
    public Builder clearVariantSetId() {
      
      variantSetId_ = getDefaultInstance().getVariantSetId();
      onChanged();
      return this;
    }
    /**
     * <code>optional string variant_set_id = 1;</code>
     *
     * <pre>
     * Required. The variant set to which variant data should be imported.
     * </pre>
     */
    public Builder setVariantSetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      variantSetId_ = value;
      onChanged();
      return this;
    }

    private com.google.protobuf.LazyStringList sourceUris_ = com.google.protobuf.LazyStringArrayList.EMPTY;
    private void ensureSourceUrisIsMutable() {
      if (!((bitField0_ & 0x00000002) == 0x00000002)) {
        sourceUris_ = new com.google.protobuf.LazyStringArrayList(sourceUris_);
        bitField0_ |= 0x00000002;
       }
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public com.google.protobuf.ProtocolStringList
        getSourceUrisList() {
      return sourceUris_.getUnmodifiableView();
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public int getSourceUrisCount() {
      return sourceUris_.size();
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public java.lang.String getSourceUris(int index) {
      return sourceUris_.get(index);
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public com.google.protobuf.ByteString
        getSourceUrisBytes(int index) {
      return sourceUris_.getByteString(index);
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public Builder setSourceUris(
        int index, java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  ensureSourceUrisIsMutable();
      sourceUris_.set(index, value);
      onChanged();
      return this;
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public Builder addSourceUris(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  ensureSourceUrisIsMutable();
      sourceUris_.add(value);
      onChanged();
      return this;
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public Builder addAllSourceUris(
        java.lang.Iterable<java.lang.String> values) {
      ensureSourceUrisIsMutable();
      com.google.protobuf.AbstractMessageLite.Builder.addAll(
          values, sourceUris_);
      onChanged();
      return this;
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public Builder clearSourceUris() {
      sourceUris_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      bitField0_ = (bitField0_ & ~0x00000002);
      onChanged();
      return this;
    }
    /**
     * <code>repeated string source_uris = 2;</code>
     *
     * <pre>
     * A list of URIs referencing variant files in Google Cloud Storage. URIs can
     * include wildcards [as described here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
     * Note that recursive wildcards ('**') are not supported.
     * </pre>
     */
    public Builder addSourceUrisBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  ensureSourceUrisIsMutable();
      sourceUris_.add(value);
      onChanged();
      return this;
    }

    private int format_ = 0;
    /**
     * <code>optional .google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
     *
     * <pre>
     * The format of the variant data being imported. If unspecified, defaults to
     * to `VCF`.
     * </pre>
     */
    public int getFormatValue() {
      return format_;
    }
    /**
     * <code>optional .google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
     *
     * <pre>
     * The format of the variant data being imported. If unspecified, defaults to
     * to `VCF`.
     * </pre>
     */
    public Builder setFormatValue(int value) {
      format_ = value;
      onChanged();
      return this;
    }
    /**
     * <code>optional .google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
     *
     * <pre>
     * The format of the variant data being imported. If unspecified, defaults to
     * to `VCF`.
     * </pre>
     */
    public com.google.genomics.v1.ImportVariantsRequest.Format getFormat() {
      com.google.genomics.v1.ImportVariantsRequest.Format result = com.google.genomics.v1.ImportVariantsRequest.Format.valueOf(format_);
      return result == null ? com.google.genomics.v1.ImportVariantsRequest.Format.UNRECOGNIZED : result;
    }
    /**
     * <code>optional .google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
     *
     * <pre>
     * The format of the variant data being imported. If unspecified, defaults to
     * to `VCF`.
     * </pre>
     */
    public Builder setFormat(com.google.genomics.v1.ImportVariantsRequest.Format value) {
      if (value == null) {
        throw new NullPointerException();
      }
      
      format_ = value.getNumber();
      onChanged();
      return this;
    }
    /**
     * <code>optional .google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
     *
     * <pre>
     * The format of the variant data being imported. If unspecified, defaults to
     * to `VCF`.
     * </pre>
     */
    public Builder clearFormat() {
      
      format_ = 0;
      onChanged();
      return this;
    }

    private boolean normalizeReferenceNames_ ;
    /**
     * <code>optional bool normalize_reference_names = 5;</code>
     *
     * <pre>
     * Convert reference names to the canonical representation.
     * hg19 haploytypes (those reference names containing "_hap")
     * are not modified in any way.
     * All other reference names are modified according to the following rules:
     * The reference name is capitalized.
     * The "chr" prefix is dropped for all autosomes and sex chromsomes.
     * For example "chr17" becomes "17" and "chrX" becomes "X".
     * All mitochondrial chromosomes ("chrM", "chrMT", etc) become "MT".
     * </pre>
     */
    public boolean getNormalizeReferenceNames() {
      return normalizeReferenceNames_;
    }
    /**
     * <code>optional bool normalize_reference_names = 5;</code>
     *
     * <pre>
     * Convert reference names to the canonical representation.
     * hg19 haploytypes (those reference names containing "_hap")
     * are not modified in any way.
     * All other reference names are modified according to the following rules:
     * The reference name is capitalized.
     * The "chr" prefix is dropped for all autosomes and sex chromsomes.
     * For example "chr17" becomes "17" and "chrX" becomes "X".
     * All mitochondrial chromosomes ("chrM", "chrMT", etc) become "MT".
     * </pre>
     */
    public Builder setNormalizeReferenceNames(boolean value) {
      
      normalizeReferenceNames_ = value;
      onChanged();
      return this;
    }
    /**
     * <code>optional bool normalize_reference_names = 5;</code>
     *
     * <pre>
     * Convert reference names to the canonical representation.
     * hg19 haploytypes (those reference names containing "_hap")
     * are not modified in any way.
     * All other reference names are modified according to the following rules:
     * The reference name is capitalized.
     * The "chr" prefix is dropped for all autosomes and sex chromsomes.
     * For example "chr17" becomes "17" and "chrX" becomes "X".
     * All mitochondrial chromosomes ("chrM", "chrMT", etc) become "MT".
     * </pre>
     */
    public Builder clearNormalizeReferenceNames() {
      
      normalizeReferenceNames_ = false;
      onChanged();
      return this;
    }
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return this;
    }

    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return this;
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.ImportVariantsRequest)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.ImportVariantsRequest)
  private static final com.google.genomics.v1.ImportVariantsRequest DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.ImportVariantsRequest();
  }

  public static com.google.genomics.v1.ImportVariantsRequest getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  public static final com.google.protobuf.Parser<ImportVariantsRequest> PARSER =
      new com.google.protobuf.AbstractParser<ImportVariantsRequest>() {
    public ImportVariantsRequest parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      try {
        return new ImportVariantsRequest(input, extensionRegistry);
      } catch (RuntimeException e) {
        if (e.getCause() instanceof
            com.google.protobuf.InvalidProtocolBufferException) {
          throw (com.google.protobuf.InvalidProtocolBufferException)
              e.getCause();
        }
        throw e;
      }
    }
  };

  public static com.google.protobuf.Parser<ImportVariantsRequest> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<ImportVariantsRequest> getParserForType() {
    return PARSER;
  }

  public com.google.genomics.v1.ImportVariantsRequest getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

